### DISCUSSION TITLE
GPIO button - sleep time?

### DISCUSSION THREAD

**Discussion by baumrasen** [photobooth-app/photobooth-app#280] [General]
I use an arcade button (60mm) with a common arcade microswitch. They are connected to the GPIO-pin and ground. They are working.

My problem is, that I get on every press action the popup with "There is already a job running...". On very fast press action the counter shows for example 14. 

I also tryed trigger on "pressend" and "released". This makes no difference.

Is there a sleep.time? Is it possible, to customize the "sleep.time"?

**Comment by mgineer85** [comment]
Unfortunately there is a bug in the underlying gpiozero implementation that prevents us from denouncing the buttons https://github.com/gpiozero/gpiozero/issues/550

Not sure yet what the best solution could look like … or the best workaround

**Reply by baumrasen** [reply]
I tryed a simple script:

```
import RPi.GPIO as GPIO
import time

GPIO.setmode(GPIO.BCM)

GPIO.setup(23, GPIO.IN, pull_up_down=GPIO.PUD_UP)

while True:
    input_state = GPIO.input(23)
    if input_state == False:
        print('Button Pressed')
        time.sleep(0.2)
```

This a this time.sleep value the message appears as expected. The lower values I got more "Button Pressed" rows at on press.

**Reply by mgineer85** [reply]
yes, but we use gpiozero instead rpi.gpio because latter is not maintained any more and doesnt support Pi5.
Also time.sleep would block calls so it needs tricks to keep the app responsive.

**Reply by baumrasen** [reply]
What, if you had a "counter-divisor". To say the app, that only use the trigger, if button is pressed "counter-divisor" times...
Or would it be possible to use the time between push and release? For example: only use events with a timewindow of 0.1-0.4 seconds between push and release?

**Reply by mgineer85** [reply]
Hey latest version has a workaround integrated. Please let me know if it works well for you as I currently do not use the GPIO :)

**Reply by baumrasen** [reply]
Hey
I tested the latest version with the buttons. It works much better then before. But on very fast push, the event is not noticed.

**Reply by mgineer85** [reply]
Denouncing is set to 60ms. You think it bothers or is just for testing so fast?

**Comment by baumrasen** [comment]
i think it's quite okay.  maybe a shorter time is also good. if I press a bit longer there is still no multiple button press count

**Reply by mgineer85** [reply]
It's not about how long the button is hold but about how bad/good the button is:
https://www.allaboutcircuits.com/technical-articles/switch-bounce-how-to-deal-with-it/

Usually it would require hardware (low pass RC filter) to debounce the buttons, but for convenience the software can do it (if it can handle it well ^^;)

**Reply by baumrasen** [reply]
Thanks for the link. I understand, that the hardware of the button can switch on and off multiple times for only on push and the debouncing do something like a real RC filter in software. 
But why the is no event fired, wenn the button is pressed very short? With my small script I see, that there are always events - also on very short press actions. 
What does [this](https://github.com/photobooth-app/photobooth-app/blob/a29929c2bcf31b80c20b6df08f464d35945ff261/photobooth/services/gpioservice.py#L137C17-L137C26) hold time do?

EDIT:
okay, I looked at the [gpiozero-documentation](https://gpiozero.readthedocs.io/en/stable/api_input.html).
But I still have no idea, why the event is not fired on very short press action :-(

**Reply by mgineer85** [reply]
Which script you refer to?

Maybe you could try to lower the value here to test which works best for you.  If you find a better value we could change that. I think 0.04 would also be sufficient if it changes the behaviour for you and still does not double execute.

 https://github.com/photobooth-app/photobooth-app/blob/b06d7a40d6bf8bfe7d69b3e1b069f7c623b5aecc/photobooth/services/gpioservice.py#L26

**Reply by baumrasen** [reply]
hold_time in line 13

**Reply by baumrasen** [reply]
I wrote a short test-script


Results:

1. DEBOUNCE_TIME = None 
    Multiple events#
1. DEBOUNCE_TIME = 0.06
    No event on very short press action
1. DEBOUNCE_TIME = 0.04
    works like a charm
1. DEBOUNCE_TIME = 0.001
    works like a charm

```
from gpiozero import Button as ZeroButton
from signal import pause

DEBOUNCE_TIME = 0.001

def say_hello():
    print("Hello!")

def say_goodbye():
    print("Goodbye!")

def say_hold():
    print("Hold!")

from gpiozero import Button as ZeroButton

class Button(ZeroButton):
    def _fire_held(self):
        # Sometimes the kernel omits edges, so if the last
        # deactivating edge is omitted held keeps firing. So
        # check the current value and send a fake edge to
        # EventsMixin to stop the held events.
        if self.value:
            super()._fire_held()
        else:
            self._fire_events(self.pin_factory.ticks(), False)

button = Button(pin=23, bounce_time=DEBOUNCE_TIME, hold_time = 0.6)

while True:
    button.when_pressed = say_hello
    button.when_released = say_goodbye
    button.when_held = say_hold
```

**Reply by mgineer85** [reply]
k, internet says typical bounce time is 20ms, so 40ms is well on the safe side. I will change it to that value in next release.
Thank you for testing!
